rust/regex-rs/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+
+                    Err(_errmsg_ostring) => Err(format!("({})", errcode)),
+                }
+            }
+        }
+    }
+
+    /// Matches input string against regex, looking for up to `max_matches` matches.
+    ///
+    /// Regexes can contain parenthesized subexpressions. This method will return up to
+    /// `max_matches`-1 of those. First match is reserved for the text that the whole regex
+    /// matched.
+    ///
+    /// `flags` dictate how matching is performed. See `MatchFlags` for details.
+    ///
+    /// # Returns
+    ///
+    /// - `Ok` with an empty vector if no match found, or if `max_matches` is 0 and there were no
+    ///   errors.
+    /// - `Ok` with a non-empty vector if `max_matches` was non-zero and a match was found. First
+    ///   element of the vector is the text that regex as a whole matched. The rest of the elements
+    ///   are pieces of text that were matched by parenthesized subexpressions.
+    /// - `Err` with an error message.
+    pub fn matches(
+        &self,
+        input: &str,
+        max_matches: usize,
+        flags: MatchFlags,
+    ) -> Result<Vec<Match>, String> {
+        let input =
+            CString::new(input).map_err(|_| String::from("Input string contains NUL byte"))?;
+
+        let mut pmatch: Vec<regmatch_t>;
+
+        let errcode = unsafe {
+            pmatch = vec![mem::zeroed(); max_matches];
+
+            regexec(
+                &self.regex,
+                input.into_raw(),
+                max_matches as libc::size_t,
+                pmatch.as_mut_ptr(),
+                flags.bits(),
+            )
+        };
+
+        match errcode {
+            0 => {
+                // Success. Let's copy results
+                let mut matches: Vec<Match> = Vec::new();
+
+                for m in pmatch {
+                    if m.rm_so < 0 || m.rm_eo < 0 {
+                        // Since `max_matches` can be bigger than the number of parenthesized
+                        // blocks in the regex, it's possible that some of the `pmatch` values are
+                        // empty. We exit the loop after detecting first such value.
+                        break;
+                    }
+
+                    // It's safe to cast i32 to usize here:
+                    // - we already checked that the values aren't negative
+                    // - usize's upper bound is higher than i32's
+                    matches.push(Match {
+                        start_pos: m.rm_so as usize,
+                        end_pos: m.rm_eo as usize,
+                    });
+                }
+
+                Ok(matches)
+            }
+
+            libc::REG_NOMATCH => {
+                // Matching went okay, but nothing found
+                Ok(Vec::new())
+            }
+
+            // POSIX only specifies two return codes for regexec(), but implementations are free to
+            // extend that.
+            _ => Err(format!("({}) match failure", errcode)),
```

Comment:
Might be nice to mention in the error message that the number is an error code (and maybe nice to mention regexec)

Reply:
Just to be clear, I don't know much about error handling in Rust so feel free to ignore.
 Especially if it is normal to have messages without context in the string.

Reply:
> Might be nice to …

Agreed. I also think I should've marked those for translation. Will fix.

> Especially if it is normal to have messages without context in the string.

What context do you mean? I don't think this error string should mention that it comes from `matches()` method, regardless of the language. The caller knows what they're calling, and if they want to propagate the error further, it's their job to add context: they might want to mention `matches()`, or maybe they'll mention the function they're calling `matches()` from, or maybe something else. Basically, it's easy for them to add to the string, but hard to remove from it.

Reply:
I rephrased the error message, marked it for translation, and ran `make extract`. Please take a look.